### PR TITLE
Inline type aliases

### PR DESCRIPTION
**Pull request by weswigham**
This introduces a new typenode with the syntax:
```
type Identifier = TypeNode
```
where within the child `TypeNode`, `Identifier` is seen as an in-scope reference to the type represented by `TypeNode`. This mostly reuses out existing type alias machinery (with no parse support for generics). This is used along the lines of:
```
interface Container {
  invoke: type TInvoke = () => TInvoke;
}
```

This allows anonymous self-referential types to rounttrip thru declaration emit. I'll be spending a bit collecting the related usecases and beefing up the tests, but the implementation is largely done, so I wanted to open a PR to 🚲 🏠 the syntax on. Notable discussion points include:

* What priority should it parse at? (currently the same as a `TypeOperator`)
* Should a sigil other than `=` be used?
* Should a keyword other than `type` be used?
* Should the syntax be expanded to allow mutually recursive aliases and/or `in` context expressions, akin to a recursive `let`?

### CODE REVIEW

**Comment by weswigham** [comment]
cc @DanielRosenwasser @uniqueiniquity and @RyanCavanaugh who briefly got to preview this earlier.

**Comment by ahejlsberg** [comment]
Without rendering an overall opinion, how about `is` instread of `=`? So:

```ts
interface Container {
  invoke: type TInvoke is () => TInvoke;
}
```

**Comment by weswigham** [comment]
I personally feel like that's a bit wordy, since a series of 3 or more non-sigil tokens starts looking more like prose than code; but that's just my own opinion - it might be that that does have nicer aesthetics. (Likewise multiple consecutive sigils is too symbolic; a alternating sequence is what usually looks normal to me)

**Comment by DanielRosenwasser** [comment]
Yeah, I'd also prefer to keep it consistent with type alias declarations unless doing so would trample on likely future syntax.

**Comment by RyanCavanaugh** [comment]
Search term: Named type expression

**Comment by SrBrahma** [comment]
+1

I come from https://github.com/microsoft/TypeScript/issues/40780. Even this not being the exact the syntax I proposed days ago, I still think this is very necessary for a better DX in more advanced types definitions. Any updates on this?

**Comment by weswigham** [comment]
> So, with the syntax I was thinking, the example in the OP might read:

Yeah, I can imagine allowing that. I'd want the comma expression to be optional, so you don't _have_ to repeat yourself in cases where it's identical like that, but it does seem useful, especially alongside the multi-let form (which would already be comma-separated).

**Comment by sandersn** [comment]
This is quite old now. I think the correct next step is to agree on a design, so I'm going to close this PR for housekeeping purposes.
However, since this has been open for 6 years with no discussion for 4, that indicates that we may not need or want it anyway.
